contracts/src/v0.8/vrf/VRFCoordinatorV2Plus.sol

```diff
@@ -0,0 +1,611 @@
+  }
+
+  /*
+   * @notice Fulfill a randomness request
+   * @param proof contains the proof and randomness
+   * @param rc request commitment pre-image, committed to at request time
+   * @return payment amount billed to the subscription
+   * @dev simulated offchain to determine if sufficient balance is present to fulfill the request
+   */
+  function fulfillRandomWords(Proof memory proof, RequestCommitment memory rc) external nonReentrant returns (uint96) {
+    uint256 startGas = gasleft();
+    Output memory output = getRandomnessFromProof(proof, rc);
+
+    uint256[] memory randomWords = new uint256[](rc.numWords);
+    for (uint256 i = 0; i < rc.numWords; i++) {
+      randomWords[i] = uint256(keccak256(abi.encode(output.randomness, i)));
+    }
+
+    delete s_requestCommitments[output.requestId];
+    VRFConsumerBaseV2 v;
+    bytes memory resp = abi.encodeWithSelector(v.rawFulfillRandomWords.selector, output.requestId, randomWords);
+    // Call with explicitly the amount of callback gas requested
+    // Important to not let them exhaust the gas budget and avoid oracle payment.
+    // Do not allow any non-view/non-pure coordinator functions to be called
+    // during the consumers callback code via reentrancyLock.
+    // Note that callWithExactGas will revert if we do not have sufficient gas
+    // to give the callee their requested amount.
+    bool success = callWithExactGas(rc.callbackGasLimit, rc.sender, resp);
+
+    // stack too deep error
+    {
+      // We want to charge users exactly for how much gas they use in their callback.
+      // The gasAfterPaymentCalculation is meant to cover these additional operations where we
+      // decrement the subscription balance and increment the oracles withdrawable balance.
+      uint96 payment = calculatePaymentAmount(
+        startGas,
+        s_config.gasAfterPaymentCalculation,
+        tx.gasprice,
+        rc.nativePayment
+      );
+      if (rc.nativePayment) {
+        if (s_subscriptions[rc.subId].ethBalance < payment) {
+          revert InsufficientBalance();
+        }
+        s_subscriptions[rc.subId].ethBalance -= payment;
+        s_withdrawableEth[s_provingKeys[output.keyHash]] += payment;
+      } else {
+        if (s_subscriptions[rc.subId].balance < payment) {
+          revert InsufficientBalance();
+        }
+        s_subscriptions[rc.subId].balance -= payment;
+        s_withdrawableTokens[s_provingKeys[output.keyHash]] += payment;
+      }
+
+      // Include payment in the event for tracking costs.
+      // event RandomWordsFulfilled(uint256 indexed requestId, uint256 outputSeed, uint96 payment, bool nativePayment, bool success);
+      emit RandomWordsFulfilled(output.requestId, output.randomness, payment, rc.nativePayment, success);
+
+      // Update the payment field in the request payments mapping so that
+      // consumers know how much they've paid and can charge end-users accordingly.
+      s_requestPayments[output.requestId] = payment;
```

Comment:
This should be a struct that indicates native/not native as well as the amount.

Reply:
Will remain 1 storage slot; however, do we even need this? Why not include this in the callback rather than storing it explicitly, and let the customer figure things out? This is like storing the request beacon output when we don't have to.

Reply:
Synced on this and it seems making it optional is a good approach. Also, it should definitely be a struct that indicates whether the payment was in native or LINK.